edgelet/edgelet-config/src/yaml_file_source.rs

```diff
@@ -0,0 +1,133 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::PathBuf;
+
+use config::{ConfigError, Source, Value};
+use yaml_rust::{Yaml, YamlLoader};
+
+/// This is similar to [`config::File`] in that it parses a YAML string or file and implements `config::Source`.
+///
+/// We use this to parse our config.yaml instead of `config::File` because `config::File` lower-cases all field names that it reads.
+/// This causes issues with fields like `agent.config.createOptions` since the config crate returns `agent.config.createoptions`
+/// which the serde deserializer ignores.
+#[derive(Clone, Debug)]
+pub(crate) enum YamlFileSource {
+    File(PathBuf),
+    String(&'static str),
+}
+
+impl Source for YamlFileSource {
+    fn clone_into_box(&self) -> Box<dyn Source + Send + Sync> {
+        Box::new(self.clone())
+    }
+
+    fn collect(&self) -> Result<HashMap<String, Value>, ConfigError> {
+        let origin = match self {
+            YamlFileSource::File(path) => Some(path.to_string_lossy().into_owned()),
+            YamlFileSource::String(_) => None,
+        };
+
+        let s = match self {
```

Comment:
nit: s is not a very readable variable name :)

Reply:
Done.